File: Leetcode/python/sol703.py
#  https://leetcode.com/problems/kth-largest-element-in-a-stream/
import heapq
class KthLargest:
    def __init__(self, k: int, nums: List[int]):
        self.arr = nums
        heapq.heapify(self.arr)
        self.kth = k

    def check_size(self) -> None:
        while len(self.arr) > self.kth:
            heapq.heappop(self.arr)

    def add(self, val: int) -> int:
        heapq.heappush(self.arr, val)
        self.check_size()
        return self.arr[0]

# 4 5 8     2 4 5 8
# 3 4 5     2 3 4 5 8
# 4 5 5     2 3 4 5 5 8
# 5 5 10    2 3 4 5 5 8 10
# 5 9 10    2 3 4 5 5 8 9 10
# 4 9 10    2 3 4 4 5 5 8 10




# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: Leetcode/cpp/sol703.cpp

// https://leetcode.com/problems/kth-largest-element-in-a-stream/description/
class KthLargest {
public:
    priority_queue<int, std::vector<int>, std::greater<int>> minIntHeap;
    int kth;
    KthLargest(int k, vector<int>& nums) {
        for(int num:nums){
            minIntHeap.push(num);
        }
        kth = k;
        check_size();
    }

    void check_size(){
        while (minIntHeap.size() > kth){
             minIntHeap.pop();
        }
    }

    int add(int val) {
        minIntHeap.push(val);
        check_size();
        return minIntHeap.top();

    }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */